python/rmm/_lib/memory_resource.pxd

```diff
@@ -111,11 +101,8 @@ cdef class PoolMemoryResource(MemoryResource):
 cdef class FixedSizeMemoryResource(MemoryResource):
     pass
 
-cdef class FixedMultiSizeMemoryResource(MemoryResource):
-    pass
-
-cdef class HybridMemoryResource(MemoryResource):
-    pass
+cdef class BinningMemoryResource(MemoryResource):
```

Comment:
We'll need to keep references to the bin resources added via `add_bin` in a Python list:

```suggestion
cdef class BinningMemoryResource(MemoryResource):
    cdef list bin_resources
```

Reply:
Do we want to use a List or something like a `std::vector` or other type that doesn't require the GIL?

Reply:
Why does Python need to track these?  Since the API that takes a resource is not exposed, all the bin resources are owned and lifetime managed by the `binning_memory_resource` on the C++ side. (`std::vector<std::unique_ptr<fixed_size_memory_resource<Upstream>>> owned_bin_resources_;`)

Reply:
Or do you intend this for exposing the other API?

Reply:
Yes, I meant this to be used by the other API which accepts a `device_memory_resource*`

Reply:
What about @kkraus14 's question?

Reply:
Instead of the `list` approach, I decided to just store a `vector` of shared pointers in the `binning_memory_resource_wrapper`. This should avoid the GIL, right?

Reply:
> This should avoid the GIL, right?

Yes -- although we aren't dropping the GIL in the `BinnedMemoryResource.add_bin()` method currently. 

> Instead of the list approach, I decided to just store a vector of shared pointers in the binning_memory_resource_wrapper. 

This LGTM. However, we sacrifice the ability to introspect/query the bin resources from Python land*. I'm not sure if any performance benefits gained from potentially avoiding the GIL here are worth it, especially given that MemoryResources in Python are only created/interacted with at startup. Thoughts?

*Actually now that I think about it a `dict` the maps allocation sizes to bin resources would have been more appropriate than a `list`.

Reply:
I don't like building duplicate and/or useful functionality in Python when we can make it available to all from C++.  For example, if you think there's a need to introspect the bins we can add that to the C++ class easy enough.

Reply:
It would be way simpler in Python just to keep references to the user-provided MemoryResources, than to make librmm API calls and construct new ones each time the user asks for them. I wouldn't say this is new functionality, rather just a bit of additional book-keeping.

Reply:
How do I make a reference in Python?

Reply:
The problem is there is no way for Python to know the type of the resource for a given bin.  If the user calls add_bin(256), then a fixed_size_memory_resource is created from the C++ side, owned by the C++ MR -- no way to access it. If the user subsequently calls add_bin(256, my_mr) then that call is a no-op since that bin already exists. 

I can add a way to query the resources for a given bin, but you will get a pointer to the base class. 

In general as we add sophisticated layered memory resources, I don't know if we want to go down the road of Python having to introspect all the layers this way.

Reply:
This is `add_bin`: 
```
cpdef add_bin(
        self,
        size_t allocation_size,
        object bin_resource=None
    ):
        cdef MemoryResource _bin_resource

        if bin_resource is None:
            (<binning_memory_resource_wrapper*>(self.c_obj.get()))[0].add_bin(
                allocation_size
            )
        else:
            # Coerce Python object `bin_resource` to C object `_bin_resource`
            _bin_resource = bin_resource

            (<binning_memory_resource_wrapper*>(self.c_obj.get()))[0].add_bin(
                allocation_size,
                _bin_resource.c_obj
            )
```

As you can see, there are two paths -- for one, Python creates the MR and therefore knows what it is. For the other, it doesn't.

Reply:
Ok yeah - good point. So I guess let's not worry about this just yet then.

Reply:
Agree we can add it if and when it is really needed.